docs/usage/writing-logic-thunks.mdx

```diff
@@ -0,0 +1,463 @@
+// standard middleware definition, with 3 nested functions:
+// 1) Accepts `{dispatch, getState}`
+// 2) Accepts `next`
+// 3) Accepts `action`
+const thunkMiddleware =
+  ({ dispatch, getState }) =>
+  next =>
+  action => {
+    // If the "action" is actually a function instead...
+    if (typeof action === 'function') {
+      // then call the function and pass `dispatch` and `getState` as arguments
+      return action(dispatch, getState)
+    }
+
+    // Otherwise, it's a normal action - send it onwards
+    return next(action)
+  }
+```
+
+In other words:
+
+- If you pass a function into `dispatch`, the thunk middleware sees that it's a function instead of an action object, intercepts it, and calls that function with `(dispatch, getState)` as its arguments
+- If it's a normal action object (or anything else), it's forwarded to the next middleware in the chain
+
+### Injecting Values Into Thunks
+
+The thunk middleware does have one customization option. You can create a custom instance of the thunk middleware at setup time, and inject an "extra argument" into the middleware. The middleware will then inject that extra value as the third argument of every thunk function. This is most commonly used for injecting an API service layer into thunk functions, so that they don't have hardcoded dependencies on the API methods:
+
+```js title="Thunk setup with an extra argument"
+import thunkMiddleware from 'redux-thunk'
+
+const serviceApi = createServiceApi('/some/url')
+
+const thunkMiddlewareWithArg = thunkMiddleware.withExtraArgument({ serviceApi })
+```
+
+Redux Toolkit's `configureStore` supports [this as part of its middleware customization in `getDefaultMiddleware`](https://redux-toolkit.js.org/api/getDefaultMiddleware):
+
+```js title="Thunk extra argument with configureStore"
+const store = configureStore({
+  reducer: rootReducer,
+  middleware: getDefaultMiddleware =>
+    getDefaultMiddleware({
+      thunk: {
+        extraArgument: { serviceApi }
+      }
+    })
+})
+```
+
+There can only be one extra argument value. If you need to pass in multiple values, pass in an object containing those.
+
+The thunk function will then receive that extra value as its third argument:
+
+```js title="Thunk function with extra argument"
+export const fetchTodoById =
+  todoId => async (dispatch, getState, extraArgument) => {
+    // In this example, the extra arg is an object with an API service inside
+    const { serviceApi } = extraArgument
+    const response = await serviceApi.getTodo(todoId)
+    dispatch(todosLoaded(response.todos))
+  }
+```
+
+## Thunk Usage Patterns
+
+### Dispatching Actions
+
+Thunks have access to the `dispatch` method. This can be used to dispatch actions, or even other thunks. This can be useful for dispatching multiple actions in a row (although [this is a pattern that should be minimized]), or orchestrating complex logic that needs to dispatch at multiple points in the process.
```

Comment:
`this is a pattern that should be minimized` is probably supposed to be a link to https://redux.js.org/style-guide/style-guide#avoid-dispatching-many-actions-sequentially

Reply:
Wondering if this might be better with an example that is more encouraged (and ties into `createAsyncThunk`)
```
dispatch({ type: "fetchData/pending" })
try {
  const results = await thing();
  dispatch({ type: "fetchData/fulfilled", payload: results })
  } catch (e) {
    dispatch({ type: "fetchData/rejected", error: e })
  }
  ```
  
  Mine is a little bit long too though, so I'm not sure. Also are async funcs ok? 🤔